### DISCUSSION TITLE
Named tags in BLOM master?

### DISCUSSION THREAD

**Discussion by TomasTorsvik** [NorESMhub/BLOM#295] [General]
Question from Mariana: 
Could we start having named tags on blom master? It would be much easier to reference these as opposed to hashes and would make this consistent with the rest of the NorESM component tagging.

**Comment by TomasTorsvik** [comment]
I think it is fine to have more tags. Any suggestions for a naming convention for the tags?

**Reply by jmaerz** [reply]
Hi @TomasTorsvik , just as a starting point: https://semver.org/ - but maybe NoRESM has its own convention?

**Reply by TomasTorsvik** [reply]
I think the "MAJOR.MINOR.PATCH" convention is useful for code releases, and this is what we follow (except we haven't used PATCH). As I understand Maraina, the question is to have other tags in BLOM that are not necessarily related to a release version, but can be referenced when doing e.g. "checkout_externals" with a modified `Externals.cfg` file.
@mvertens - maybe you can explain better how you see the use of these tags?

**Comment by mvertens** [comment]
Lets say we have a blom tag naming convention on master as: blom_x.y.z (you already have v1.3.0 - so we could keep that as well).
Every commit to master would update the z now. And then you would decide when you wanted to update y and x for some of the commits to master. 
The updating of z can be done automatically in github actions. See
https://github.com/NorESMhub/WW3/blob/noresm/.github/workflows/bumpversion.yml

**Reply by TomasTorsvik** [reply]
OK, this seems like a reasonable solution to me. 
@matsbn , @JorgSchwinger , @jmaerz , @milicak 
do you have any opinion regarding this way of making tags on the `master` branch?

**Reply by jmaerz** [reply]
I don't have a strong opinion on this - the advantage that I see is that the tagging allows to better identify the time line and getting a grasp of code change from the tag, while sha would be shorter to type, but being rather random and without the additional meaning.

**Comment by TomasTorsvik** [comment]
@mvertens 
I see one potential drawback with tagging every commit in `master`. When I look at the commit log I frequently use `git log --oneline --decorate --graph --simplify-by-decoration` (or the `emacs magit` equivalent to this command) to get a log with only the last commits on each branch. But tags are also "decorated" in the log, so with automatic tagging it is less straight forwad to hide "intermediate" commits on the `master` branch. I'm not aware of any solution where you can either hide all tags or only include the head of each branch. Do you know if this is possible somehow?
This is perhaps only an issue for me, but it would be great to have this functionality available.

**Reply by gold2718** [reply]
When I try that command on a Blom clone:
```
git log --oneline --decorate --graph --simplify-by-decoration origin/master
```
I get:
```
* 0b07349 (origin/master, origin/HEAD) Move main sediment parameters to mo_param_bgc (#297)
* 1d71682 (master) Merge pull request #281 from mvertens/feature/refactor_ifdefs3
* e343fee (origin/feature-langmuir) Added diagnostics of wave fields received from CESM coupler.
*   0749d29 Merge branch 'master' into feature-hamocc-vsls-upd
|\  
* | dd8bbe8 (origin/feature-hamocc-vsls) Merge pull request #93 from tjiputra/feature-hamocc-vsls
| * 0c7f492 (tag: v1.1.0, origin/release-1.1) Merge pull request #76 from AleksiNummelin/feature_highres
|/  
* a8a7e94 (tag: pre-v1.0.0) Merge pull request #24 from DirkOlivie/master
* 4c83041 (tag: blom0_0_001) Merge pull request #10 from matsbn/feature-inputdata
* 56d027d (tag: blom0_0_000) Merge pull request #6 from matsbn/feature-blom-lgpl
* 8d1b58b Initial import of structure. Code is based on ccsm4_0_alpha31
```
Is that the output you see that you would like changed? If not, what am I doing wrong in my command line?

**Reply by TomasTorsvik** [reply]
Yes, that is the output I get as well, and this behavior is what I want to preserve.
The problem with having automated tagging is that you will get [tag: blom*.*.*] all over the master, so the branch heads will get lost in a long list of tags, i.e. `--simplify-by-decoration` will not simplify if everything is decorated.
There may be a solution with adding `--decorate-refs-exclude='/refs/tags/blom*`, which seem to work on the command line, but I haven't found a `magit` equivalent (https://git-scm.com/docs/git-log)

**Reply by TomasTorsvik** [reply]
OK, it seems it should be possible to exclude the tags from the log after all, so this is probably a non-issue.

**Comment by jmaerz** [comment]
To bring the discussion here: 
Jörgs comment (https://github.com/NorESMhub/BLOM/pull/297#issuecomment-1790532220)

> 
> Regarding the tags: What would be the relation between these automated tags and our releases? We have currently tagged releases v1.0.0 to v1.3.0 and we are working towards v1.4.0 (the last CMIP6 backwards compatible release). So I would find it confusing to have the automated tag v1.0.0 associated with our release v1.4.0 - is there a reasonable way to synchronize this?
> 

Then it would be at least logical to start with `blom_v1.4.0`.

**Reply by TomasTorsvik** [reply]
I agree, we should have some consistency in numbering between the automated tags and last release version.

**Comment by matsbn** [comment]
I can see the benefit of more tags. As I understand this would be for tagging development snapshots, not releases, so I think we should not directly increment on the v[MAJOR].[MINOR].[PATCH] tag names we use for releases (although we have no patch releases yet, as pointed out above). Maybe a tag name containing "dev" in it? I'm no fan of having the component name as a prefix of a tag name, as it just seems redundant because a tag without a reference to the repository is meaningless.

Great if we can wait to decide on this until after 15 Nov :)

**Comment by mvertens** [comment]
Waiting until after 15 Nov makes sounds fine. Thanks for all of the input on this topic.

**Comment by mvertens** [comment]
The recommendation from the meeting as I understand it:
releases: v1.5.0 , v1.5.1 (bug fixes) .... v1.6.0
development: dev1.5.0.1, dev1.5.0.2, (automated by github actions) ..... dev1.6.0.1

As an example:
we tag v1.5.0 now
we start release testing and there is a problem
we create a release branch and update the code and tag it as v1.5.1

**Comment by TomasTorsvik** [comment]
I have created `dev1.4.0.1` on the `master` branch now.

**Reply by jmaerz** [reply]
@matsbn and @TomasTorsvik: Can someone please summarize what has been decided on today? - I eventually got lost and still somewhat have the question: if the release version number `v1.X.y` shall be reflected in the `master`  `dev1.X.y.m` -tag, then I wonder why answer changing `dev`-tags can have only an incremental `m` change (the current case between `v1.4.0`  - assuming this being based on a currently non-tagged `dev1.4.0.0` -  and now the answer changing `master` `dev1.4.0.1`). I am essentially asking, how we see answer-changing code merges in `master` reflected in the `dev` tags - also against the background of the release `v` versions?

**Reply by JorgSchwinger** [reply]
We have had a strong focus on bit-for-bit compatibility up to now to maintain compatibility with cmip6, but in general I think it is more the rule that a code change also will be answer-changing? So not sure if we need/want to reflect this in the version tags? Might complicate things.

**Reply by jmaerz** [reply]
Hi @JorgSchwinger , I guess, I just try to understand the reasoning at the moment. Wouldn't we want to see answer changes reflected in the tags in `master`? Technically, this could be achieved by perceiving the first two digits belonging to the clock-giving release version and the digit 3 for answer changes (or bigger changes) and digit 4 for minor, non-answer changes (all this for `master`). In the release version, one either has no answer changes or one documents those (and the third digit would reflect code changes - answer changing or not). Thus far, I don't see much release tag change going beyond the second digit (so rather `v1.1.0`, `v1.2.0`, etc. - so why would we want to keep the third digit also for the `master` dev-tag? - also, if the tag `dev1.4.0.1` isn't compatible any further with `v1.4.0`? - then it would be easier to just increment the third digit...). I guess, I will be fine with everything at the end, I just would like to know how it is handled and explicitly written out (as in which number corresponds to what). I attach a figure to illustrate, how I would/could perceive it after our discussions (and orient along #164) while I understand that it is likely not the way how it was discussed today.
 
![git-flow-to-share](https://github.com/NorESMhub/BLOM/assets/92309038/2e91ac4b-ae76-4edc-89ba-2652e575f656)

So having a `dev1.4.a.p` and parallel a `v1.4.x` (technically, a similar `v1.4.a'.p'` could be used to signal changes in answer in the release version...). But maybe it's too late to consider this and I will be fine. In any case, it would be good to document the new convention somewhere.

**Reply by TomasTorsvik** [reply]
Hi @jmaerz , @matsbn , @JorgSchwinger , @mvertens ,
thanks for drawing this out. I think this is what we more or less agreed on yesterday, i.e. that `devX.Y.Z.n` follows `master` and `vX.Y.Z` indicates releases. So far so good.
The purpose of introducing `dev` tags in the first place is to have an easy naming convention to refer to in the NorESM `Externals.cfg` file, instead of referring to commit hashes. I don't think we should attach too much meaning to the numbering, I just see it as e.g. "dev1.4.0.1 follows after v1.4.0".
It's probably correct that we will rarely, if ever, increment the 3rd number in the `dev` tag. One possible scenario where this could happen is if, let's say we have v1.5.0 as latest release, and there are new machine settings that are introduced in both master and the latest release version, we may move to a `dev1.5.1.n` along side `v1.5.1`. Bu I think this will be the exceptional case rather than the rule.
I suggest we try this out for now. If we see that it needs adjusting, maybe we can target a new major release, e.g. v2.0.0, and make adjustments then. I think it will be easier to identify what works and what doesn't when we use this system and get some experience with it.

**Reply by jmaerz** [reply]
Dear @TomasTorsvik , many thanks for clarification - so just to state that clearly: the last digit is just an incrementer without further meaning apart from stating another commit. Further, any third digit increment `Z` in `dev1.4.Z.n` also doesn't have any further meaning besides that is tells, whether any `v1.4.Z` was merged back into `master` (as in they can totally deviate in terms of answers, since code in `master` can have progressed differently). A bit confusing, but thanks for clarification. I really appreciate your feedback!

**Reply by TomasTorsvik** [reply]
@jmaerz , great, I think we are on the same page here. Hope we have everyone else on board as well. I think this will become clearer once we start using this system, or perhaps, we will see clearer if there are any pitfalls.

**Reply by mvertens** [reply]
@TomasTorsvik - thanks for adding the clarifications. 
@jmaerz @matsbn @JorgSchwinger - I implemented the new github actions for automatic tagging - but after talking to @TomasTorsvik this morning we feel that we should defer the automatic tagging for now until we are all comfortable with the new tagging scheme. I can easily do  PR with the new automatic scheme once that is the case.

**Reply by TomasTorsvik** [reply]
@mvertens - Thanks!
I suggest we include the `dev` tags "by hand" until the `v1.5.0` tag, which probably will come next week, and make that a target starting point for automatic tagging, i.e. to be introduced as `dev1.5.0.1`. Is that OK with everyone?

**Reply by jmaerz** [reply]
@TomasTorsvik basically like this:

![git-workflow-crop](https://github.com/NorESMhub/BLOM/assets/92309038/8e494116-3248-4f11-92a8-db17eded92e3)

**Reply by jmaerz** [reply]
Dear @TomasTorsvik, @matsbn, @mvertens and @JorgSchwinger , I guess, during and after the discussion, I was/am a bit perplexed that the new system shouldn't hold any information on code changes in the development `master` branch. Flagging for answer changes would also enforce the regression testing a bit. But I do see that the maintenance costs and overhead are probably too high for this, so just go ahead from my side.

**Comment by TomasTorsvik** [comment]
@jmaerz @matsbn @JorgSchwinger @mvertens @milicak 
Hi, I have continued to tag occasionally "by hand" according to the scheme outlined in the figure above, but not for every merge commit on `master`. I don't see any problems with having more tags, but it is a bit annoying that github always produce tar and zip archives for each tag. The size of these archives are about 1.6MB in total, so it will take a while before this becomes a real issue.

I still wonder if we need to tag everything, but it would be the easiest and most consistent way to proceed. I also see that the tags may become more useful in the development phase towards NorESM2.5 and NorESM3, if we make more rapid changes in the Externals.cfg file.
